src/components/SignTransactionForm.tsx

```diff
@@ -0,0 +1,187 @@
+                className="button is-fullwidth"
+                onClick={this.signTransaction}
+                disabled={!this.state.transaction || this.state.signed}
+              >
+                Sign
+              </button>
+            </div>
+            <div className="control is-expanded">
+              <button
+                className="button is-fullwidth"
+                onClick={this.copyTransaction}
+                disabled={!this.state.transaction || !this.state.signed}
+              >
+                Copy
+              </button>
+            </div>
+            <div className="control is-expanded">
+              <button
+                className={`button is-fullwidth ${
+                  this.props.submissionPending ? 'is-loading' : ''
+                }`}
+                onClick={this.submitTransaction}
+                disabled={!this.state.signed || this.props.submissionPending}
+              >
+                Submit
+              </button>
+            </div>
+          </div>
+        </div>
+        {this.state.transaction && (
+          <TransactionView
+            transaction={this.state.transaction}
+            connection={this.props.connection}
+            didSign={this.state.signed}
+          />
+        )}
+      </div>
+    )
+  }
+}
+
+interface Props {
+  transaction: Transaction
+  connection: Connection
+  didSign: boolean
+}
+interface TransactionState {
+  signers: string[]
+}
+class TransactionView extends React.Component<Props, TransactionState> {
+  state: TransactionState = { signers: [] }
+
+  getSigners = async () => {
+    const signers = await getTransactionSigners(
+      getServer(this.props.connection),
+      this.props.transaction,
+    )
+    this.setState({ signers: signers.map(sig => sig.publicKey()) })
+  }
+
+  componentDidMount() {
+    this.getSigners()
+  }
+
+  componentDidUpdate(prevProps: Props) {
+    if (prevProps.didSign !== this.props.didSign) {
+      this.getSigners()
+    }
+  }
+
+  renderOperations = () =>
+    this.props.transaction.operations.map(op => <OperationOverview operation={op} key={op.type} />)
+
+  renderSigners = () =>
+    this.state.signers.map(sig => <HorizontalLabelledField key={sig} value={sig} label="" />)
+
+  render() {
+    const { transaction } = this.props
+    return (
+      <div className="section">
+        <h3 className="subtitle">Transaction</h3>
+        <HorizontalLabelledField label="Source" value={transaction.source} />
+        <HorizontalLabelledField label="Fee" value={transaction.fee.toString()} />
+        {this.renderOperations()}
+        <hr />
+        <h3 className="subtitle">Signers</h3>
+        {this.renderSigners()}
+      </div>
+    )
+  }
+}
+
+class OperationOverview extends React.Component<{ operation: TransactionOperation }> {
+  renderOperationRecords = () =>
+    Object.entries(this.props.operation)
+      .filter(([_, value]) => typeof value === 'string' || typeof value === 'number')
```

Comment:
these filter calls seem like they might read better the other way, what I mean is:

```js
.filter(([_, value]) => ['string','number'].includes(typeof value)) // wrap test values in array
.filter(([key]) => key !== 'type') // unwrap test key array
```

in which case we could do it at the same time without too much bloat?

```js
.filter(([key, value]) => key !== 'type && ['string','number'].includes(typeof value))
```

Reply:
I like it